From e0380b9adb5f1c101d174d724e06f0b899836e8b Mon Sep 17 00:00:00 2001
Message-ID: <e0380b9adb5f1c101d174d724e06f0b899836e8b.1750071117.git.maciej.borzecki@canonical.com>
From: Maciej Borzecki <maciej.borzecki@canonical.com>
Date: Thu, 5 Sep 2024 14:06:50 +0200
Subject: [PATCH] Build position-independent binaries per hardening policy

This is only supported on some architectures and only wth recent glibc (where rcrt1.o is present).

Signed-off-by: Maciej Borzecki <maciej.borzecki@canonical.com>
---
 cmd/Makefile.am    | 4 ++--
 packaging/snapd.mk | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/cmd/Makefile.am b/cmd/Makefile.am
index 3a64b145084a1d440d741fa704c8ef2af14cef91..1945600ab33200f3674c2a168c377f9de732321d 100644
--- a/cmd/Makefile.am
+++ b/cmd/Makefile.am
@@ -515,7 +515,7 @@ snap_gdb_shim_snap_gdb_shim_SOURCES = \
 	snap-gdb-shim/snap-gdb-shim.c
 
 snap_gdb_shim_snap_gdb_shim_LDADD = libsnap-confine-private.a
-snap_gdb_shim_snap_gdb_shim_LDFLAGS = -static
+snap_gdb_shim_snap_gdb_shim_LDFLAGS = -static-pie
 
 ##
 ## snap-gdbserver-shim
@@ -531,7 +531,7 @@ snap_gdb_shim_snap_gdbserver_shim_SOURCES = \
 	snap-gdb-shim/snap-gdbserver-shim.c
 
 snap_gdb_shim_snap_gdbserver_shim_LDADD = libsnap-confine-private.a
-snap_gdb_shim_snap_gdbserver_shim_LDFLAGS = -static
+snap_gdb_shim_snap_gdbserver_shim_LDFLAGS = -static-pie
 
 ##
 ## snapd-generator
diff --git a/packaging/snapd.mk b/packaging/snapd.mk
index bf8efeb176c7060b238416df8112fac74f1593e5..e3a3fa6b034172f581cd5a74839d2a9aad05bf65 100644
--- a/packaging/snapd.mk
+++ b/packaging/snapd.mk
@@ -70,7 +70,7 @@ GO_MOD=-mod=readonly
 endif
 
 # Go -ldflags settings for static build. Can be overridden in snapd.defines.mk.
-EXTRA_GO_STATIC_LDFLAGS ?= -linkmode external -extldflags="-static" $(EXTRA_GO_LDFLAGS)
+EXTRA_GO_STATIC_LDFLAGS ?= -linkmode external -extldflags="-static-pie" $(EXTRA_GO_LDFLAGS)
 
 # NOTE: This *depends* on building out of tree. Some of the built binaries
 # conflict with directory names in the tree.
@@ -93,7 +93,7 @@ $(builddir)/snap $(builddir)/snap-seccomp $(builddir)/snapd-apparmor:
 # nearly-arbitrary mount namespace that does not contain anything we can depend
 # on (no standard library, for example).
 $(builddir)/snap-update-ns $(builddir)/snap-exec $(builddir)/snapctl:
-	go build -o $@ -buildmode=default \
+	go build -o $@ -buildmode=pie \
 		$(GO_MOD) \
 		$(if $(GO_TAGS),-tags "$(GO_TAGS)") \
 		-ldflags="$(EXTRA_GO_STATIC_LDFLAGS)" \
-- 
2.49.0

